File: warranty-service/src/service/WarrantyVerificationService.ts
import { CoveragePeriod } from '../domain/CoveragePeriod.js';
import { CoverageScope } from '../domain/CoverageScope.js';
import { DefectCategory, parseDefectCategory } from '../domain/DefectCategory.js';
import { Warranty } from '../domain/Warranty.js';
import { WarrantyRepository } from '../repository/WarrantyRepository.js';
import { coerceToUuid } from '../util/idCoerce.js';

export interface WarrantyRegisteredEvent {
  contractId: string;
  unitId: string;
  customerId: string;
  startsAt: string;
  endsAt: string;
  coveredCategories?: string[];
}

export interface DefectReportedEvent {
  defectId: string;
  contractId: string;
  unitId: string;
  customerId: string;
  defectCategory: string;
  description: string;
  reportedAt?: string;
}

export interface WarrantyVerifiedEvent {
  claimId: string;
  warrantyId: string;
  defectId: string;
  contractId: string;
  unitId: string;
  customerId: string;
  coverageStatus: string;
  coverageReason?: string;
  verifiedAt?: string;
  expiresAt: string;
}

const ONE_YEAR_SECONDS = 365 * 24 * 3600;

export const WarrantyVerificationService = {
  async registerWarranty(event: WarrantyRegisteredEvent): Promise<Warranty> {
    console.log(`[Command] RegisterWarranty — contractId=${event.contractId}, unitId=${event.unitId}`);

    const period = new CoveragePeriod(event.startsAt, event.endsAt);
    const categories = (event.coveredCategories ?? Object.values(DefectCategory)).map((c) =>
      parseDefectCategory(c),
    );
    const scope = new CoverageScope(categories);

    const warranty = Warranty.register({
      contractId: coerceToUuid(event.contractId),
      unitId: coerceToUuid(event.unitId),
      customerId: coerceToUuid(event.customerId),
      period,
      scope,
    });
    await WarrantyRepository.save(warranty);

    console.log(`[Domain Event] WarrantyRegistered: warrantyId=${warranty.warrantyId}`);
    return warranty;
  },

  async verifyDefect(event: DefectReportedEvent): Promise<WarrantyVerifiedEvent> {
    console.log(`[Command] VerifyClaim — defectId=${event.defectId}, contractId=${event.contractId}`);

    // Coerce string codes (e.g. "PROP-001") to deterministic UUIDs
    const contractId = coerceToUuid(event.contractId);
    const unitId = coerceToUuid(event.unitId);
    const customerId = coerceToUuid(event.customerId);
    const defectId = coerceToUuid(event.defectId);

    const reportedAt = event.reportedAt ?? new Date().toISOString();
    const reportedTime = new Date(reportedAt);

    let warranty = await WarrantyRepository.findByContractId(contractId);
    if (!warranty) {
      console.warn(
        `[Service] No warranty found for contractId=${contractId}, creating default 1y full-scope warranty`,
      );
      // Anchor coverage on reportedAt (with a small buffer) so a defect that
      // arrives slightly before "now" doesn't fall outside the period due to
      // clock skew between Post-sale and Legal.
      const anchor = new Date(Math.min(reportedTime.getTime(), Date.now()) - 60_000);
      const ends = new Date(anchor.getTime() + ONE_YEAR_SECONDS * 1000);
      const period = new CoveragePeriod(anchor.toISOString(), ends.toISOString());
      const scope = new CoverageScope(Object.values(DefectCategory));
      warranty = Warranty.register({
        contractId,
        unitId,
        customerId,
        period,
        scope,
      });
      await WarrantyRepository.save(warranty);
    }

    const category = parseDefectCategory(event.defectCategory);

    const claim = warranty.verifyClaim({
      defectId,
      defectCategory: category,
      description: event.description,
      reportedAt,
    });
    await WarrantyRepository.save(warranty);

    console.log(
      `[Domain Event] WarrantyVerified: claimId=${claim.claimId}, status=${claim.coverageStatus}, reason=${claim.coverageReason}`,
    );

    return {
      claimId: claim.claimId,
      warrantyId: warranty.warrantyId,
      defectId: claim.defectId,
      contractId: warranty.contractId,
      unitId: warranty.unitId,
      customerId: warranty.customerId,
      coverageStatus: claim.coverageStatus,
      coverageReason: claim.coverageReason,
      verifiedAt: claim.verifiedAt,
      expiresAt: warranty.coveragePeriod.endsAt,
    };
  },
};

File: _java-archive/warranty-service/src/main/java/dev/kritchalach/warranty/service/WarrantyVerificationService.java
package dev.kritchalach.warranty.service;

import dev.kritchalach.warranty.domain.*;
import dev.kritchalach.warranty.event.DefectReportedEvent;
import dev.kritchalach.warranty.event.WarrantyRegisteredEvent;
import dev.kritchalach.warranty.event.WarrantyVerifiedEvent;
import dev.kritchalach.warranty.repository.WarrantyRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * Application Service สำหรับ Flow 4: Defect Report
 * - รับ warranty.registered → สร้าง Warranty aggregate
 * - รับ defect.reported → ตรวจสอบ + emit warranty.verified
 */
@Service
public class WarrantyVerificationService {

    private static final Logger log = LoggerFactory.getLogger(WarrantyVerificationService.class);

    private final WarrantyRepository warrantyRepository;

    public WarrantyVerificationService(WarrantyRepository warrantyRepository) {
        this.warrantyRepository = warrantyRepository;
    }

    /**
     * Command: RegisterWarranty
     */
    public Warranty registerWarranty(WarrantyRegisteredEvent event) {
        UUID contractId = UUID.fromString(event.getContractId());
        UUID unitId = UUID.fromString(event.getUnitId());
        UUID customerId = UUID.fromString(event.getCustomerId());

        log.info("Command: RegisterWarranty — contractId={}, unitId={}", contractId, unitId);

        CoveragePeriod period = new CoveragePeriod(
                Instant.parse(event.getStartsAt()),
                Instant.parse(event.getEndsAt()));

        Set<DefectCategory> categories = event.getCoveredCategories() != null
                ? Arrays.stream(event.getCoveredCategories())
                        .map(DefectCategory::valueOf)
                        .collect(Collectors.toCollection(HashSet::new))
                : new HashSet<>(Arrays.asList(DefectCategory.values()));
        CoverageScope scope = new CoverageScope(categories);

        Warranty warranty = Warranty.register(contractId, unitId, customerId, period, scope);
        warranty = warrantyRepository.save(warranty);

        log.info("Warranty registered: warrantyId={}", warranty.getWarrantyId());
        return warranty;
    }

    /**
     * Command: VerifyClaim
     * ถ้ายังไม่มี Warranty registered จะสร้าง default warranty (1 ปี เต็ม scope) ให้
     * เพื่อให้ระบบยังคง process defect ได้แม้ Post-sale ลืมส่ง warranty.registered มาก่อน
     */
    public WarrantyVerifiedEvent verifyDefect(DefectReportedEvent event) {
        UUID defectId = UUID.fromString(event.getDefectId());
        UUID contractId = UUID.fromString(event.getContractId());
        UUID unitId = UUID.fromString(event.getUnitId());
        UUID customerId = UUID.fromString(event.getCustomerId());

        log.info("Command: VerifyClaim — defectId={}, contractId={}", defectId, contractId);

        Warranty warranty = warrantyRepository.findByContractId(contractId)
                .orElseGet(() -> {
                    log.warn("No warranty found for contractId={}, creating default 1y full-scope warranty",
                            contractId);
                    Instant now = Instant.now();
                    CoveragePeriod period = new CoveragePeriod(now, now.plusSeconds(365L * 24 * 3600));
                    CoverageScope scope = new CoverageScope(
                            new HashSet<>(Arrays.asList(DefectCategory.values())));
                    Warranty fresh = Warranty.register(contractId, unitId, customerId, period, scope);
                    return warrantyRepository.save(fresh);
                });

        DefectCategory category = parseCategory(event.getDefectCategory());
        Instant reportedAt = event.getReportedAt() != null
                ? Instant.parse(event.getReportedAt())
                : Instant.now();

        WarrantyClaim claim = warranty.verifyClaim(defectId, category,
                event.getDescription(), reportedAt);
        warranty = warrantyRepository.save(warranty);

        log.info("Claim verified: claimId={}, status={}, reason={}",
                claim.getClaimId(), claim.getCoverageStatus(), claim.getCoverageReason());

        return new WarrantyVerifiedEvent(
                claim.getClaimId(),
                warranty.getWarrantyId(),
                claim.getDefectId(),
                warranty.getContractId(),
                warranty.getUnitId(),
                warranty.getCustomerId(),
                claim.getCoverageStatus().name(),
                claim.getCoverageReason(),
                claim.getVerifiedAt(),
                warranty.getCoveragePeriod().getEndsAt()
        );
    }

    private DefectCategory parseCategory(String raw) {
        if (raw == null) return DefectCategory.OTHER;
        try {
            return DefectCategory.valueOf(raw.toUpperCase());
        } catch (IllegalArgumentException e) {
            return DefectCategory.OTHER;
        }
    }
}
